src/DiffResults_helpers.jl

```diff
@@ -14,7 +14,7 @@ Allocate a DiffResults buffer for a gradient, taking the element type of `x` int
 function _diffresults_buffer(ℓ, x)
     T = eltype(x)
     S = T <: Real ? float(Real) : Float64 # heuristic
-    DiffResults.MutableDiffResult(zero(S), (Vector{S}(undef, dimension(ℓ)), ))
+    DiffResults.MutableDiffResult(zero(S), (similar(x, S), ))
```

Comment:
Should this still take into account `dimension`? I'm not sure but I assume there was a reason why it was used instead of `x`?
```suggestion
    DiffResults.MutableDiffResult(zero(S), (similar(x, S, dimension(ℓ)), ))
```

Reply:
>  using the dimension from x

So `similar` basically does this. And AFAIK this is only called from the `logdensity_and_gradient` method, meaning that `x` should always have the correct dimensions.

Using `dimension` from `ℓ` instead will sometimes result in loss of structure of the original type for `x`, e.g. `ComponentArrays`.

Reply:
I would recommend something like
```julia
function _diffresults_buffer(x)
    T = eltype(x)
    S = T <: Real ? float(Real) : Float64 # heuristic
    DiffResults.MutableDiffResult(zero(S), (similar(x, S), ))
end
```
with the caller modified accordingly.

Reply:
This is what I've impleented, right? Or are my eyes deceiving me?

Reply:
I think @tpapp's point is that the first argument of the function should be dropped if we do not use it anymore.